{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python version 3.4.3 |Anaconda 2.3.0 (x86_64)| (default, Mar  6 2015, 12:07:41) \n",
      "[GCC 4.2.1 (Apple Inc. build 5577)]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Hi baby, welcome to SynBioCAD.\n",
      "Refer to this community google doc for documentation:\n",
      "https://docs.google.com/document/d/11YNoUBtit4rkWZQr_ZclzAOTE_HGMtgB3H_6cvq1Y24/edit?usp=sharing\n",
      "\n",
      "\n",
      "What do you want to do today?\n",
      "\n",
      "\t 1 \t add DNA into a characterized locus\n",
      "\n",
      "\t 2 \t edit an existing gene, e.g., delete or replace\n",
      "\n",
      "\t 3 \t build a custom cassette\n",
      "\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "SynBioCAD\n",
    "by Leo d'Espaux and William Zhuang\n",
    "\n",
    "Current as of 1/12/2016\n",
    "\n",
    "Recent updates:\n",
    "*Added figures with current locus and promoter level\n",
    "*Added image of CNE1 tag, to expand on later\n",
    "'''\n",
    "\n",
    "\n",
    "# import libraries we're using\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "import copy\n",
    "from intermine.webservice import Service\n",
    "from pandas import *\n",
    "from pandas import DataFrame, read_csv\n",
    "import pandas as pd  \n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "from IPython.display import Image\n",
    "from IPython.display import display\n",
    "%matplotlib inline \n",
    "\n",
    "\n",
    "\n",
    "# define global variables\n",
    "HomologyLength = 1000\n",
    "PrimerMaxTm = 55 # As calculated by statluc, more like 65C by Thermo's Phusion program\n",
    "PrimerMaxLen = 60\n",
    "OverhangMaxFrac = 1\n",
    "\n",
    "\n",
    "\n",
    "def askUser():\n",
    "    print('Python version ' + sys.version+\"\\n\\n\\n\\n\")\n",
    "    print(\"Hi baby, welcome to SynBioCAD.\")\n",
    "    print(\"Refer to this community google doc for documentation:\")\n",
    "    print(\"https://docs.google.com/document/d/11YNoUBtit4rkWZQr_ZclzAOTE_HGMtgB3H_6cvq1Y24/edit?usp=sharing\")\n",
    "    print(\"\\n\\nWhat do you want to do today?\\n\")\n",
    "    print(\"\\t 1 \\t add DNA into a characterized locus\\n\") \n",
    "    print(\"\\t 2 \\t edit an existing gene, e.g., delete or replace\\n\")\n",
    "    print(\"\\t 3 \\t build a custom cassette\\n\")\n",
    "\n",
    "   \n",
    "    Action= input(\"Your answer: \")\n",
    "    if Action == \"1\":\n",
    "        editEmpty()\n",
    "        \n",
    "    elif Action == \"2\":\n",
    "        editExisting()\n",
    "\n",
    "    elif Action == \"3\": # This is more for stitching things together not necessarily for CRISPR\n",
    "        buildCustom()\n",
    "             \n",
    "            \n",
    "            \n",
    "def editEmpty():    \n",
    "    \n",
    "    # I have an excel file where we keep cut site information\n",
    "    # and load it as a pandas dataframe called cutFrame\n",
    "    \n",
    "    cutFrame = pd.read_excel(\"locusTable.xlsx\", index_col=\"cutName\")\n",
    "    \n",
    "    print(\"\\nThese are the sites we have:\\t\")\n",
    "    \n",
    "    LocusPic=Image(\"LocusLevelExp.png\")\n",
    "    display(LocusPic)\n",
    " \n",
    "    print(cutFrame)\n",
    "         \n",
    "    print(\"\\nWhich cut site do you want, e.g., ARS208a?\\n\")\n",
    "    cutName=input(\"Your answer: \")\n",
    "\n",
    "    # cutName is a string that should correspond to one of the variables\n",
    "    # of the \"cutName\" column in dataFrame, e.g., ARS208a\n",
    "    \n",
    "    # cutFrame has the 20nt sequence of the gRNA saved, which we load\n",
    "    # It also has some more information we'll get to later\n",
    "    cutSequence=cutFrame.loc[cutName,'cutSequence']\n",
    "\n",
    "    # We need to load the chromosome where the cut site is\n",
    "    location=cutFrame.loc[cutName,'cutChr']+\".fasta\"\n",
    "     \n",
    "    ChromosomeSeq=SeqIO.read(location, \"fasta\").seq\n",
    "    \n",
    "    if ChromosomeSeq.find(cutSequence)==-1:\n",
    "        ChromosomeSeq=ChromosomeSeq.reverse_complement()\n",
    "        if ChromosomeSeq.find(cutSequence)==-1:\n",
    "            print(\"CAN'T FIND THE CUT SITE IN YOUR SEQUENCE\")\n",
    "\n",
    "    StartIndex=ChromosomeSeq.find(cutSequence)\n",
    "    EndIndex=StartIndex+34\n",
    "    \n",
    "    UpSeq=ChromosomeSeq[StartIndex-HomologyLength:StartIndex]\n",
    "    DownSeq=ChromosomeSeq[EndIndex:EndIndex+HomologyLength]\n",
    "        \n",
    "    UpHomRec = SeqRecord(UpSeq, name=\"up\"+cutName)\n",
    "    DownHomRec = SeqRecord(DownSeq, name=cutName)\n",
    "    \n",
    "    print(\"\\nFound your homology regions. What do you want to do to \" + cutName + \"?\\n\")\n",
    "    print(\"\\t1\\t You have an ORF but want help picking promoter and terminator\\n\")\n",
    "    print(\"\\t2\\t You already have a one-piece cassette and just want to insert it\\n\\n\")\n",
    "    # Coming later\n",
    "    # print(\"\\t3\\t You want to construct a donor containing two standard cassettes (pointing away from each other\\n\")\n",
    "\n",
    "    typeEdit=input(\"Your answer: \")\n",
    "    if typeEdit==\"1\":\n",
    "    \n",
    "        PromoterRec, orfRecord, TerminatorRec = buildCassette()\n",
    "        \n",
    "        fragments=[UpHomRec, PromoterRec, orfRecord, TerminatorRec, DownHomRec]\n",
    "\n",
    "    \n",
    "    elif typeEdit == \"2\":\n",
    "        \n",
    "        orfName=input(\"What's the name of your custom gene or cassette?\")\n",
    "        orfSeq=Seq(input(\"What's the sequence?\"))\n",
    "        orfRecord=SeqRecord(orfSeq,name=orfName)\n",
    "        \n",
    "        fragments=[UpHomRec, orfRecord, DownHomRec]\n",
    "\n",
    "    elif typeEdit == \"3\":\n",
    "        \n",
    "        print(\"Your construct will look like this: \")\n",
    "        print(\"up hom-<terminator1-gene1-promoter1<->promoter2-gene2-terminator2>-downhom\")\n",
    "        print(\"OK, let's build your FIRST cassette\")\n",
    "        PromoterRec1, orfRecord1, TerminatorRec1 = buildCassette()\n",
    "         \n",
    "        #now in the construction, these will be in the antisense direction, so let's change their sequences\n",
    "        rTerminatorRec1=flipRecord(TerminatorRec1)\n",
    "        rorfRecord1=flipRecord(orfRecord1)\n",
    "        rPromoterRec1=flipRecord(PromoterRec1)\n",
    "        \n",
    "        print(\"OK, let's build your SECOND cassette\")\n",
    "        PromoterRec2, orfRecord2, TerminatorRec2 = buildCassette()\n",
    "\n",
    "        fragments=[UpHomRec, rTerminatorRec1, rorfRecord1, rPromoterRec1,PromoterRec2,orfRecord2, TerminatorRec2,DownHomRec]\n",
    "\n",
    "    stitch(fragments)\n",
    "    \n",
    "    \n",
    "    \n",
    "def editExisting():\n",
    "    \n",
    "    print(\" \")\n",
    "    print(\"Which locus do you want to edit? Tell me a common name, e.g., \\\"OAF1\\\": \")\n",
    "    print(\"I'm smart and pretty and I can fetch it for you.\\n\")\n",
    "    \n",
    "    GeneName= input(\"Your answer: \")\n",
    "    \n",
    "    OrigGeneRecord=fetchGene(GeneName)\n",
    "    # note that this returns a seqrecord\n",
    "    \n",
    "    print(\"\\nOK, I found \"+OrigGeneRecord.features+\":\")\n",
    "    print(OrigGeneRecord.description)\n",
    "\n",
    "    \n",
    "    # We make seqrecords since that's what we carry through later in the program\n",
    "\n",
    "    UpHomRec = fetchNeighbor(OrigGeneRecord, \"upstream\", HomologyLength )\n",
    "    UpHomRec.name=UpHomRec.name+\"up\"\n",
    "    DownHomRec = fetchNeighbor(OrigGeneRecord, \"downstream\", HomologyLength )\n",
    "    DownHomRec.name=DownHomRec.name+\"down\"\n",
    "\n",
    "    print(\"\\nWhat do you want to do to this gene?\\n\")\n",
    "    print(\"\\t1\\t delete the protein coding DNA sequence (CDS)\\n\")\n",
    "    print(\"\\t2\\t replace the CDS with a cassette--I'll help you pick Promoter and Terminator\\n\")\n",
    "    print(\"\\t3\\t replace the CDS with another CDS\\n\")\n",
    "    #print(\"\\t4\\t replace the CDS with a custom cassette (soon)\")\n",
    "    #print(\"\\t5\\t replace a specified region near your target gene (soon)\")\n",
    "    \n",
    "    Action=input(\"Your answer: \")\n",
    "                        \n",
    "    #note that in all the below, we want to have fragments be records\n",
    "    if Action==\"1\":\n",
    "        fragments=[UpHomRec,DownHomRec]\n",
    "    \n",
    "    if Action==\"2\":\n",
    "        PromoterRec, orfRecord, TerminatorRec = buildCassette()\n",
    "        fragments=[UpHomRec, PromoterRec, orfRecord, TerminatorRec, DownHomRec] #we need to finish buildcassette to add InsertRec here\n",
    "\n",
    "    if Action==\"3\":\n",
    "        print(\" \")\n",
    "        NewGeneName=input(\"What's the name of the gene you're inserting?\")\n",
    "        NewGeneSeq=Seq(input(\"What's the sequence of your new gene?\"))\n",
    "        InsertRec = SeqRecord(NewGeneSeq, name=NewGeneName)\n",
    "        fragments=[UpHomRec, InsertRec, DownHomRec]\n",
    "\n",
    "    if Action==\"4\":\n",
    "        print(\"How many pieces (other than homology fragments) are you stitching together.\")\n",
    "        Npieces=input(\"Your answer: \")\n",
    "        output = buildCassette()\n",
    "\n",
    "        fragments=[UpHomRec, DownHomRec] #we need to finish buildcassette to add InsertRec here\n",
    "\n",
    "    # Now let's pick the cut site. This is independent from the donor cassette\n",
    "    # We have an excel sheet with genes and pre-picked cut sites\n",
    "    # Or let the user pick\n",
    "    \n",
    "    print(\"\\nIf you have a 20-nt cut sequence enter it now. Otherwise, type \\\"p\\\" and I'll find one.\\n\")\n",
    "    \n",
    "    cutChoice=input(\"Your answer: \")\n",
    "    \n",
    "    if cutChoice==\"p\":\n",
    "        cutSequence=cutFinder(GeneName)\n",
    "    elif len(cutChoice)==20:\n",
    "        cutSequence=cutChoice\n",
    "        # Later should check that the sequence is present and unique\n",
    "    \n",
    "   \n",
    "    print(\"\\n\\n\\nHere is the oligo to clone or gap-repair your cut plasmid:\\n\")\n",
    "\n",
    "    print(\"cut\"+GeneName+\" tggcgaatgggacttt\"+cutSequence+\"gttttagagctagaaatagcaagt\\n\")\n",
    "\n",
    "    # Now we feed the design to a tool to pick primers which amplify each fragment\n",
    "    # and add terminal homology to the neighboring fragments\n",
    "    stitch(fragments)\n",
    "\n",
    "    \n",
    "    \n",
    "def buildCustom():\n",
    "    N = int(input(\"How many pieces in your custom cassette: \"))\n",
    "    fragments = variableCassette(N)[0]\n",
    "    stitch(fragments)\n",
    "\n",
    "            \n",
    "    #------------------------------ FETCH FUNCTIONS -------------------------------------\n",
    "\n",
    "    \n",
    "def fetchGene(GeneName):\n",
    "    \n",
    "    service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
    "    template = service.get_template('Gene_GenomicDNA')\n",
    "\n",
    "    rows = template.rows(\n",
    "        E = {\"op\": \"LOOKUP\", \"value\": GeneName, \"extra_value\": \"S. cerevisiae\"}\n",
    "    )\n",
    "    \n",
    "    # this service seems to return multiple similar genes but we want the first one only, so count\n",
    "    # and it returns information about the gene you want\n",
    "    count=0\n",
    "    for row in rows:\n",
    "        \n",
    "        count=count+1\n",
    "        if count==1:\n",
    "            descr= row[\"description\"]\n",
    "            GeneSeq=Seq(row[\"sequence.residues\"])\n",
    "            GeneSysName=row[\"secondaryIdentifier\"]\n",
    "       \n",
    "    #let's create a record for the oldGene\n",
    "    GeneRecord = SeqRecord(GeneSeq, id=GeneSysName)\n",
    "    \n",
    "    #now let's add some more information to make it useful\n",
    "    GeneRecord.name=GeneName\n",
    "    GeneRecord.features=GeneSysName\n",
    "    GeneRecord.description=descr\n",
    "\n",
    "    return GeneRecord \n",
    "       \n",
    "    \n",
    "    \n",
    "def fetchNeighbor(NeighborRecord, direction, distance):\n",
    "\n",
    "    # let's load the appropriate chromosome file. The record of the gene we looked up\n",
    "    # contains in the \"features\" the systematic name, wherein the second letter\n",
    "    # corresponds to chromosome number, e.g., 1=A etc\n",
    "    if NeighborRecord.features[1]==\"A\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer01.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"B\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer02.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"C\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer03.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"D\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer04.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"E\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer05.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"F\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer06.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"G\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer07.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"H\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer08.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"I\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer09.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"J\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer10.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"K\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer11.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"L\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer12.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"M\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer13.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"N\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer14.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"O\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer15.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"P\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer16.fasta\", \"fasta\") \n",
    "\n",
    "    \n",
    "    \n",
    "    # let's explicitely name the sequences from the seq record\n",
    "    NeighborSeq=NeighborRecord.seq\n",
    "    ChromosomeSeq=ChromosomeRec.seq\n",
    "    \n",
    "    # flip the sequence to orient with respect to the old gene\n",
    "    if ChromosomeSeq.find(NeighborSeq)==-1:\n",
    "        ChromosomeSeq=ChromosomeSeq.reverse_complement()\n",
    "\n",
    "    StartIndex=ChromosomeSeq.find(NeighborSeq)\n",
    "    EndIndex=StartIndex+len(NeighborSeq)\n",
    "    \n",
    "    if direction==\"upstream\":\n",
    "        DesiredSeq=ChromosomeSeq[StartIndex-distance:StartIndex]\n",
    "    if direction==\"downstream\":\n",
    "        DesiredSeq=ChromosomeSeq[EndIndex:EndIndex+distance]\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    NeighborRec = SeqRecord(DesiredSeq, name=NeighborRecord.name)\n",
    "    \n",
    "    return NeighborRec\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    #------------------------------------ CONSTRUCTING STUFF --------------------------------------\n",
    "\n",
    "def getPrimer(currRecord):\n",
    "    \n",
    "\n",
    "    mp = 0\n",
    "    length = 0\n",
    "    primer = Seq(\"\")\n",
    "\n",
    "    seq=currRecord.seq\n",
    "    \n",
    "    while mp <= PrimerMaxTm and length <= PrimerMaxLen:\n",
    "        primer = primer + seq[length]\n",
    "        mp = MeltingTemp.Tm_staluc(primer)\n",
    "        length += 1\n",
    "\n",
    "    return primer           \n",
    "        \n",
    "def overhangPrimer(currRecord,prevSeq):\n",
    "    #let's get the template-binding primer first\n",
    "    primer=getPrimer(currRecord)\n",
    "    \n",
    "    \n",
    "    #OK let's work on the overhang\n",
    "    maxOhLen=PrimerMaxLen-len(primer)    \n",
    "    maxFrac=1\n",
    "    \n",
    "    #let's decide on a max overhang length\n",
    "    if round(len(primer)*(OverhangMaxFrac+1)) < 60:\n",
    "             maxOhLen=round(len(primer)*OverhangMaxFrac)\n",
    "    \n",
    "    #the index must be an integer!!!\n",
    "    maxOhLen=int(maxOhLen)\n",
    "    ohprimer=prevSeq.seq[-maxOhLen:]+primer #we add the .seq so that it returns a string\n",
    "    \n",
    "    return ohprimer      \n",
    "        \n",
    "\n",
    "          \n",
    "def buildCassette():\n",
    "    \n",
    "    #first, the promoter\n",
    "    print(\"I'm going to build a standard cassette in which promoter is 600nt, terminator 250nt.\") \n",
    "    print(\"Here are the promoters we have characterized:\")\n",
    "    PromPic=Image(\"PromoterLevelExp.png\")\n",
    "    display(PromPic)    \n",
    "    print(\"Which PROMOTER do you want to use, e.g., \\\"TDH3\\\"\")\n",
    "    \n",
    "    PromoterName=input(\"Your answer: \")\n",
    "    PromoterGeneRec=fetchGene(PromoterName)\n",
    "    PromoterRec=fetchNeighbor(PromoterGeneRec,\"upstream\",600)\n",
    "    PromoterRec.name=PromoterRec.name+\"ps\"\n",
    "    \n",
    "    \n",
    "    #second, the terminator\n",
    "    print(\"\\nI haven't tested terminators yet, but there are good references out there, e.g., .\")\n",
    "    print(\"Which TERMINATOR do you want to use, e.g., \\\"ADH1\\\"\")\n",
    "    TerminatorName = input('Your answer: ')\n",
    "    TerminatorGeneRec=fetchGene(TerminatorName)\n",
    "    TerminatorRec=fetchNeighbor(TerminatorGeneRec,\"downstream\",250)\n",
    "    TerminatorRec.name=TerminatorRec.name+\"ts\"\n",
    "    \n",
    "    #and last, the gene\n",
    "    print(\"\\nWhat is the name of your CDS, e.g., KlGapDH\")\n",
    "    orfName = input(\"Your answer: \")\n",
    "    \n",
    "    print(\"What's the sequence\")\n",
    "    orfSeq=input(\"Your answer: \")\n",
    "    \n",
    "    orfRecord=SeqRecord(Seq(orfSeq), name=orfName)\n",
    "    \n",
    "    #print(\"\\nDo you want to attach any protein tags?\\n\")\n",
    "    #print(\"This is under construction but here is tag CNE1 targeting to the ER lumen.\\n\")\n",
    "    #CNE1tag=Image(\"CNE1tag.png\")\n",
    "    #display(CNE1tag)\n",
    "    \n",
    "    insertRec=[PromoterRec,orfRecord,TerminatorRec]\n",
    "    return PromoterRec, orfRecord, TerminatorRec\n",
    "    \n",
    "\n",
    "\n",
    "def stitch(fragments):\n",
    "    #this function takes seq records and prints primers\n",
    "    \n",
    "    #let's make an empty sequence file\n",
    "    Nfrags=len(fragments)\n",
    "    donor=Seq(\"\")\n",
    "    index=[]\n",
    "\n",
    "    print(\"\\nHere are the primers to amplify your fragments and construct your donor DNA cassette:\\n\")\n",
    "\n",
    "    for i in range (0, Nfrags):\n",
    "        donor=donor+fragments[i]\n",
    "    \n",
    "    # The names include information on the homology provided by the overhang\n",
    "    # Note that some primers don't have overhangs\n",
    "    for i in range (0, Nfrags):\n",
    "        if i==0:\n",
    "            print(\"L\"+ fragments[i].name + \" \" + getPrimer(donor))\n",
    "            print(\"R\"+ fragments[i].name + \"(\" + fragments[i+1].name + \") \" + overhangPrimer(fragments[i].reverse_complement(),fragments[i+1].reverse_complement()))\n",
    "        elif i==Nfrags-1:\n",
    "            print(\"L\"+ fragments[i].name + \"(\" + fragments[i-1].name + \") \" + overhangPrimer(fragments[i],fragments[i-1]))\n",
    "            print(\"R\"+ fragments[i].name + \" \" + getPrimer(donor.reverse_complement()))\n",
    "        else:\n",
    "            print(\"L\"+ fragments[i].name + \"(\" + fragments[i-1].name + \") \" + overhangPrimer(fragments[i],fragments[i-1]))\n",
    "            print(\"R\"+ fragments[i].name + \"(\" + fragments[i+1].name + \") \" + overhangPrimer(fragments[i].reverse_complement(),fragments[i+1].reverse_complement()))\n",
    "\n",
    "    print(\"\\n\\nThe sequence of your donor DNA--and final chromosomal sequence--is below.\")\n",
    "    print(\"Refer to the following link for more help.\")\n",
    "    print(\"https://docs.google.com/document/d/11YNoUBtit4rkWZQr_ZclzAOTE_HGMtgB3H_6cvq1Y24/edit?usp=sharing\\n\")\n",
    "\n",
    "    print(len(donor.seq))\n",
    "\n",
    "    print(donor.seq)\n",
    "\n",
    "    print(\"\\nYou might want to copy this entire prompt and save it for your records.\\n\")\n",
    "    \n",
    "        \n",
    "def flipRecord(origRecord):\n",
    "    origRecord.seq=origRecord.seq.reverse_complement()\n",
    "    origRecord.name=\"r\"+origRecord.name\n",
    "    return origRecord\n",
    "    \n",
    "def cutFinder(GeneName):\n",
    "    cutFrame = pd.read_excel(\"cutsTable.xlsx\", index_col=\"cutName\")\n",
    "    \n",
    "    # cutFrame has the 20nt sequence of the gRNA saved, which we load\n",
    "    # It also has some more information we'll get to later\n",
    "    \n",
    "    cutSequence=cutFrame.loc[GeneName,'cutSequence']\n",
    "    \n",
    "    print(\"\\nI picked the following cut site for you \"+cutSequence)\n",
    "    \n",
    "    return cutSequence\n",
    "\n",
    "\n",
    "askUser()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Raw Cell Format",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
